lib/Arch/Sleigh/ControlFlowStructuring.cpp

```diff
@@ -0,0 +1,470 @@
+#include "ControlFlowStructuring.h"
+
+#include <glog/logging.h>
+
+namespace remill::sleigh {
+
+bool isVarnodeInConstantSpace(VarnodeData vnode) {
+  auto spc = vnode.getAddr().getSpace();
+  return spc->constant_space_index == spc->getIndex();
+}
+
+
+namespace {
+
+// variant casting taken c&ped.
+template <class... Args>
+struct variant_cast_proxy {
+  std::variant<Args...> v;
+
+  template <class... ToArgs>
+  operator std::variant<ToArgs...>() const {
+    return std::visit([](auto &&arg) -> std::variant<ToArgs...> { return arg; },
+                      v);
+  }
+};
+
+template <class... Args>
+auto variant_cast(const std::variant<Args...> &v)
+    -> variant_cast_proxy<Args...> {
+  return {v};
+}
+
+enum CoarseEffect { ABNORMAL, NORMAL };
+
+struct CoarseFlow {
+  CoarseEffect eff;
+  bool is_conditional;
+};
+
+enum CoarseCategory { CAT_NORMAL, CAT_ABNORMAL, CAT_CONDITIONAL_ABNORMAL };
+
+
+static std::optional<CoarseFlow>
+CoarseFlowFromControlFlowOp(const RemillPcodeOp &op, uint64_t next_pc) {
+  if (op.op == CPUI_CALL || op.op == CPUI_CALLIND || op.op == CPUI_BRANCHIND ||
+      op.op == CPUI_RETURN) {
+    return {{CoarseEffect::ABNORMAL, false}};
+  }
+
+  // either a branch or a cbranch
+
+  // figure out if this is a fallthrough, input 0 is the next target
+
+  auto is_conditional = op.op == CPUI_CBRANCH;
+  if (isVarnodeInConstantSpace(op.vars[0])) {
+    // this is an internal branch.. we cant handle that right now
+    return std::nullopt;
+  }
+
+  if (op.vars[0].offset == next_pc) {
+    return {{CoarseEffect::NORMAL, is_conditional}};
+  } else {
+    return {{CoarseEffect::ABNORMAL, is_conditional}};
+  }
+}
+
+// gets a list of indeces and coarse categories in this pcodeop block
+static std::optional<std::map<size_t, CoarseFlow>>
+CoarseFlows(const std::vector<RemillPcodeOp> &ops, uint64_t next_pc) {
+  std::map<size_t, CoarseFlow> res;
+  size_t ind = 0;
+  for (auto op : ops) {
+    if (ControlFlowStructureAnalysis::isControlFlowPcodeOp(op.op)) {
+      auto cc = CoarseFlowFromControlFlowOp(op, next_pc);
+      if (!cc) {
+        return std::nullopt;
+      }
+
+      res.emplace(ind, *cc);
+    }
+
+    ind++;
+  }
+
+  auto last_index = ops.size();
+  // insert a pseudo control flow op at the end
+  // add a fallthrough insn at +1 to represent a last fallthrough if there is a chance we fallthrough at the end
+  if (ops.empty() ||
```

Comment:
need to check empty first, cant extract into variables since we cant access ops.size()-1 if !ops.empty(), could invert the condition then nest two ifs and return out, but that's kinda ugly

Reply:
I meant more like
```
auto good_name = whole condition;
if ( good_name ) {
  ...
}
```
or I would just move the `{` to the new line, so it is really easy to see where the condition ends.